### PR TITLE
[LOOP-2385] Use pub/sub to get the current glucose unit when updated

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-2385

The intent of this update is to allow any displayed CGM settings to call the function that gets the current preferred glucose unit (say they are monitoring `HKUserPreferencesDidChange` to detect a preference change). Without this when the unit is changed the CGM settings need to back all the way out before the updated glucose unit can be displayed.

What do you think?

### CODE REVIEW

**Review by ps2 — APPROVED** [review]
LGTM!

**Review by darinkrauss — APPROVED** [review]
I *think* this is okay, but it does assume that the UI will automatically refresh (and thus call this function) whenever the units are changed. Consider a notification function on `CGMManager` along the lines of `unitsDidChange` to more proactively and explicitly tell the manager that the units changed.

(I can think of one, admittedly extreme, edge case where this won't work involving two phones connected to the same iCloud/Health data with one displaying the Loop CGMManager UI and the other changing the units in Health.)

Approved if you and @ps2 are okay with this.

**Comment by nhamming** [comment]
@darinkrauss @ps2 I pushed up this PR before our discussion at dev sync. Currently I'm working on creating a pub/sub for glucose unit changes and when the structure of that is ready for an additional review, I'll ping you both again. 

Thanks for your input.

**Comment by nhamming** [comment]
As discussed, this is a notification based alternative to using a current glucose function. Looking for your critical thought/feedback about this approach.

Thanks!

@darinkrauss this is still at the `CGMManagerUI` layer. There is need for something similar at the `CGMManager` layer and I'm still considering this. Do you think this pattern would also work well there?

File: LoopKitUI/CGMManagerUI.swift
Comment (darinkrauss) [inline LoopKitUI/CGMManagerUI.swift:16]: [nit] Any need to be notified of glucose units changes during setup? (I suspect so.)

**Review by darinkrauss — APPROVED** [review]
Just one question, but otherwise LGTM! 🚀

File: LoopKitUI/CGMManagerUI.swift
Comment (nhamming) [reply LoopKitUI/CGMManagerUI.swift:16]: I was going to add this as well, but as you can see, the glucose units are not passed to the setup view controller. So currently there isn't a use case for this.

Comment (darinkrauss) [reply LoopKitUI/CGMManagerUI.swift:16]: I think as part of the general protocol of CGM setup/settings it would need to be there, but I suppose we can defer until later.

File: LoopKitUI/HUDProvider.swift
Comment (ps2) [inline LoopKitUI/HUDProvider.swift]: I think it would be better to not require CGMViewController to conform to GlucoseUnitObserver, but rather make it optional, and the receiver of this action can test via typecasting to see if the VC conforms.

Comment (darinkrauss) [reply LoopKitUI/HUDProvider.swift]: My concern with this is that it isn't at all clear which objects support `GlucoseUnitObserver`. Any optional check for a protocol conformance is hidden from the API contract. A comment would help, but comments go stale (and we'd need to remember to put it whenever there is optional protocol conformance).

Comment (nhamming) [reply LoopKitUI/HUDProvider.swift]: updated as suggested

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: As I commented in the `Loop` PR, couldn't we walk the cache and see specifically *what* has changed? If it isn't anything in the cache, then we don't care about it, so we can ignore it. (Well, at least I'm assuming that all preferred units in Loop/LoopKit go through the unit cache.)

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: [nit] It might make more sense to have the object be the unit cache if that is possible?

**Review by darinkrauss — COMMENTED** [review]
Just going to leave my comments first so see what you and @ps2 think.

File: LoopKit/DeviceManager/CGMManager.swift
Comment (ps2) [inline LoopKit/DeviceManager/CGMManager.swift]: I don't think this should be part of the core DeviceManager layer. Glucose units are a user *display* preference, and part of the UI.

File: LoopKit/GlucoseUnitPublishing.swift
Comment (ps2) [inline LoopKit/GlucoseUnitPublishing.swift]: I wonder if we should change this name to something like "preferredGlucoseDisplayUnitObserver", just to keep us in the mindset that this does not represent a change in storage units, but rather how the user wants to see things. I wouldn't hold up this PR for this naming change, but it's a conflation I keep seeing.

**Review by ps2 — CHANGES_REQUESTED** [review]
I think preferred glucose display units should not be in the core device manager layer, unless I'm missing something and they really do need to be there.

File: LoopKit/DeviceManager/CGMManager.swift
Comment (nhamming) [reply LoopKit/DeviceManager/CGMManager.swift]: The Dexcom alerting needs the units to display a glucose value in the alert. Originally I only conformed the Dexcom CGM manager to `GlucoseUnitObserver` and checked for conformation after the CGM manager was initialized to add it to a unit publisher. @darinkrauss suggested making the conformance mandatory. 

I'll ping you to chat about this.

Comment (rickpasetto) [reply LoopKit/DeviceManager/CGMManager.swift]: If we should be moving the `AlertPresenter` and `AlertResponder` protocol conformance out of `DeviceManager` and to `DeviceManagerUI`, I would be in favor of that, if we can make it work.

File: LoopKit/GlucoseUnitPublishing.swift
Comment (rickpasetto) [inline LoopKit/GlucoseUnitPublishing.swift]: I'm wondering if we should be hand-rolling a special-case pub/sub system for this, as opposed to using [Combine](https://developer.apple.com/documentation/combine) now that it is part of Swift.  Thoughts, @ps2 ?

**Review by rickpasetto — COMMENTED** [review]
Added some comments for thought/discussion...

**Comment by nhamming** [comment]
@ps2 updated as discussed. Let me know if I overlooked anything.

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift:34]: [nit] `.global()`?

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: [nit] `@discardableResult` and ditch the `_ = ` below?

Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: [nit] `@discardableResult` and ditch the `_ = ` below?

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: [nit] Just `return getHealthStoreUnitAndUpdateCache(for: quantityTypeIdentifier)`?

File: LoopKit/HealthStoreUnitCache.swift
Comment (darinkrauss) [inline LoopKit/HealthStoreUnitCache.swift]: Consider moving this inside the `mutate` below so there isn't the timing gap between this check and the update.

File: LoopKit/QuantityFormatter.swift
Comment (darinkrauss) [inline LoopKit/QuantityFormatter.swift:170]: What actual values are used for `mmol/L/min`? 

IIRC, it is `2` or `3` `mg/dL/min`. Doing the simple math conversion, this is `0.11101496` and `0.16652244` `mmol/L/min`. Does Dexcom round this to `0.1` and `0.2`?

Just wondering if `1` `preferredFractionDigits` is enough for `mmol/L/min`.

File: MockKitUI/View Controllers/MockCGMManagerSettingsNavigationController.swift
Comment (darinkrauss) [inline MockKitUI/View Controllers/MockCGMManagerSettingsNavigationController.swift]: For my work, I had to create a `CGMManagerSettingsNavigationController` and `PumpManagerSettingsNavigationController` for my work for other reasons (I needed to add a different delegate protocol to conform to). 

I would suggest renaming this to just `CGMManagerSettingsNavigationController`, make it a generic with a constraint that the `rootViewController` type must conform to `PreferredGlucoseUnitObserver`, make it public (like `SettingsNavigationViewController` is), and use it here and throughout the plugins. (Therefore, no need to have a separate `*SettingsNavigationViewController` for each plugin.)

**Review by darinkrauss — APPROVED** [review]
A few questions and comments and one recommendation. Approved, if you choose to pass on my recommendation. 🔭

File: LoopKit/QuantityFormatter.swift
Comment (nhamming) [reply LoopKit/QuantityFormatter.swift:170]: Yes, Dexcom G6 rounds this to 0.1 or 0.2 mmol/L/min (at least at the time when it was implemented).

File: LoopKit/HealthStoreUnitCache.swift
Comment (ps2) [inline LoopKit/HealthStoreUnitCache.swift]: nit: delete line 101, this line, and 110, 114, and put lines 111-113 here?

File: LoopKit/QuantityFormatter.swift
Comment (ps2) [inline LoopKit/QuantityFormatter.swift:149]: nit: this appears to be just a helper for tests, and could be moved there.

File: LoopKitUI/CGMManagerSettingsNavigationController.swift
Comment (ps2) [inline LoopKitUI/CGMManagerSettingsNavigationController.swift:13]: nit: this doesn't feel like a pattern worth extracting, but rather just boilerplate that happens to be reused across some of the current cgm managers, and could very well be implemented differently in a different cgm manager.

Comment (ps2) [reply LoopKitUI/CGMManagerSettingsNavigationController.swift:13]: I'd recommend keeping this in MockKitUI

**Review by ps2 — APPROVED** [review]
Generally looks ok. A few nits.

File: LoopKitUI/CGMManagerSettingsNavigationController.swift
Comment (darinkrauss) [reply LoopKitUI/CGMManagerSettingsNavigationController.swift:13]: I asked Nate to do this because for my work needs a specific `SettingsNavigationController` for cgms and one for pumps. We can, of course, implement this plus my work in a `SettingsNavigationController` in each plugin that is specific to that plugin, but if that is the case why do we even have a common `SettingsNavigationController`? Should we do away with the common `SettingsNavigationController` altogether and push into the plugin?

File: LoopKitUI/GlucoseUnitPublishing.swift
Comment (ps2) [inline LoopKitUI/GlucoseUnitPublishing.swift]: No clients are consuming this protocol, there are just classes declaring conformance to it, and I don't understand a need for this protocol.

**Review by ps2 — COMMENTED** [review]
Another comment after looking at the Loop PR. I think this should probably be simplified a bit.

File: LoopKitUI/GlucoseUnitPublishing.swift
Comment (nhamming) [reply LoopKitUI/GlucoseUnitPublishing.swift]: fair point.

**Pull request merged**
